File: src/UnwindMC/Decompilation/Decompiler.fs
module Decompiler

let decompile (project : DecompilationProject.T): unit =
    let pe = PEFile.load project.exePath
    let importResolver = new ImportResolver.ImportResolver(pe.imageBase, pe.getImportAddressTableBytes(), pe.getImportBytes())
    let analyzer = Analyzer.create (pe.getTextBytes()) pe.textSectionAddress importResolver
    Analyzer.addFunction analyzer pe.entryPointAddress
    Analyzer.analyze analyzer
    FIXME "implement the rest of the project decompilation"

let decompileFunction (graph: InstructionGraph.T) (address: uint64): string =
    let il = ILDecompiler.decompile graph address
    let blocks = FlowAnalyzer.buildFlowGraph(il)
    let result = TypeResolver.resolveTypes(blocks)
    let func = AstBuilder.buildAst (sprintf "sub_%06x" address) blocks result.parameterTypes result.localTypes result.variableTypes
    CppEmitter.emit func

File: lib/src/decompiler.rs
use asm::InstructionGraph;
use ast_builder::AstBuilder;
use cpp_emitter::CppEmitter;
use flow_analyzer::build_flow_graph;
use il_decompiler::decompile;
use type_resolver::TypeResolver;

pub fn decompile_function(graph: &InstructionGraph, address: u64) -> String {
    let il = decompile(graph, address).expect("Failed to decompile IL");
    let blocks = build_flow_graph(il);
    let (blocks, types) = TypeResolver::resolve_types(blocks);
    let func = AstBuilder::build_ast(format!("sub_{0:06x}", address), &blocks, &types);
    CppEmitter::emit(&func)
}
